dubbo-common/src/main/java/org/apache/dubbo/rpc/model/ReflectionServiceDescriptor.java

```diff
@@ -89,6 +98,21 @@ private void initMethods() {
                     +" stream method signatures. method(" + methodName + ")");
         });
     }
+    
+    public static boolean isProtobufClass(Class<?> clazz) {
+        while (clazz != Object.class && clazz != null) {
+            Class<?>[] interfaces = clazz.getInterfaces();
+            if (interfaces.length > 0) {
+                for (Class<?> clazzInterface : interfaces) {
+                    if (PROTOBUF_MESSAGE_CLASS_NAME.equalsIgnoreCase(clazzInterface.getName())) {
+                        return true;
+                    }
+                }
+            }
+            clazz = clazz.getSuperclass();
```

Comment:
Is it really in need to check super class?

Reply:
this method just be copied from ReflectionPackableMethod.java,Is there a better way to judge pb class?

Reply:
I'm worried about whether judging every time will affect performance when init.

Reply:
I understand your concern, if I keep the code logic I submitted unchanged, I can think some solutions:
1. Modify the isProtobufClass method, Judge pb class by isAssignableFrom. This method requires that the Dubco-common project add the dependency of protobuf.
2. Add cache for isProtobufClass method implementation , Cache the pb class while judging the pb class, It can reduce the judgment of the same pb class. In addition, I found that the method also be called in the rpc process, By caching pb Class,performance of rpc call can be improved by caching pb class also.
3. Adding caching in mode 1 has the same benefits as in mode 2

Reply:
I think this only happens when the service is started and registered, not when rpc is called. The performance impact can be ignored.

Reply:
> I think this only happens when the service is started and registered, not when rpc is called. The performance impact can be ignored.

The usage time during startup is also important. When the service scale is large, startup can take tens of minutes or even hours.

Reply:
> > I think this only happens when the service is started and registered, not when rpc is called. The performance impact can be ignored.
> 
> The usage time during startup is also important. When the service scale is large, startup can take tens of minutes or even hours.

What can i do fix it.

Reply:
I've written some other implementations of the isProtobufClass method. And attach some test results
`
import com.google.protobuf.Message;

import org.apache.dubbo.demo.consumer.ApiConsumer;
import org.apache.dubbo.demo.hello.HelloRequest;
import org.apache.dubbo.rpc.protocol.tri.call.TripleClientCall;

import static org.apache.dubbo.common.constants.CommonConstants.PROTOBUF_MESSAGE_CLASS_NAME;

/**
 * @author zhouning
 * @date 2023/10/07 17:07
 */
public class Test {

    public static void main(String[] args) throws NoSuchMethodException {
        System.out.println("####class is pb class start");
        long start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass(HelloRequest.class);
        }
        System.out.println("isProtobufClass cost=" + (System.currentTimeMillis() - start));
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass2(HelloRequest.class);
        }
        System.out.println("isProtobufClass2 cost=" + (System.currentTimeMillis() - start));
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass3(HelloRequest.class);
        }
        System.out.println("isProtobufClass3 cost=" + (System.currentTimeMillis() - start));
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass4(HelloRequest.class);
        }
        System.out.println("isProtobufClass4 cost=" + (System.currentTimeMillis() - start));
        System.out.println("####class is pb class end");

        System.out.println("####class is not pb class start");
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass(TripleClientCall.class);
        }
        System.out.println("isProtobufClass cost=" + (System.currentTimeMillis() - start));
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass2(TripleClientCall.class);
        }
        System.out.println("isProtobufClass2 cost=" + (System.currentTimeMillis() - start));
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass3(TripleClientCall.class);
        }
        System.out.println("isProtobufClass3 cost=" + (System.currentTimeMillis() - start));
        start = System.currentTimeMillis();
        for (int i = 0; i < 100000; i++) {
            isProtobufClass4(TripleClientCall.class);
        }
        System.out.println("isProtobufClass4 cost=" + (System.currentTimeMillis() - start));
        System.out.println("####class is not pb class end");
    }


    public static boolean isProtobufClass(Class<?> clazz) {
        while (clazz != Object.class && clazz != null) {
            Class<?>[] interfaces = clazz.getInterfaces();
            if (interfaces.length > 0) {
                for (Class<?> clazzInterface : interfaces) {
                    if (PROTOBUF_MESSAGE_CLASS_NAME.equalsIgnoreCase(clazzInterface.getName())) {
                        return true;
                    }
                }
            }
            clazz = clazz.getSuperclass();
        }
        return false;
    }

    public static boolean isProtobufClass2(Class<?> clazz) {
        try {
            //pb class must deserilizable by parseFrom(byte[]) method
            HelloRequest.class.getDeclaredMethod("parseFrom", byte[].class);
        } catch (Exception e) {
            return false;
        }
        while (clazz != Object.class && clazz != null) {
            Class<?>[] interfaces = clazz.getInterfaces();
            if (interfaces.length > 0) {
                for (Class<?> clazzInterface : interfaces) {
                    if (PROTOBUF_MESSAGE_CLASS_NAME.equalsIgnoreCase(clazzInterface.getName())) {
                        return true;
                    }
                }
            }
            clazz = clazz.getSuperclass();
        }
        return false;
    }

    public static boolean isProtobufClass3(Class<?> pojoClazz) {
        try {
            Class<?> protobufBaseClass = Class.forName(PROTOBUF_MESSAGE_CLASS_NAME);
            return protobufBaseClass.isAssignableFrom(pojoClazz);
        } catch (ClassNotFoundException | NoClassDefFoundError e) {
            // If sink does not have protobuf in classpath then it cannot be protobuf
            return false;
        }
    }

    public static boolean isProtobufClass4(Class<?> pojoClazz) {
        return Message.class.isAssignableFrom(pojoClazz);
    }
}

`
####class is pb class start
isProtobufClass cost=35
isProtobufClass2 cost=74
isProtobufClass3 cost=55
isProtobufClass4 cost=4
####class is pb class end
####class is not pb class start
isProtobufClass cost=11
isProtobufClass2 cost=46
isProtobufClass3 cost=55
isProtobufClass4 cost=4
####class is not pb class end

Reply:
There is a better implementation:
1. Use Dubbo’s Classloader in Dubbo’s tool class in advance to try to load the class name of Protobuf Message to obtain whether protobuf exists and the corresponding `Class` object
2. When judgment is needed, judge in advance whether protobuf exists. If it exists, use `Class` to directly judge `isAssignableFrom`.

Reply:
I think I understand what you say, and just to make sure, if can't load Protobuf Message class, the isProtobuf method can directly return false. After all, if you can't get Protobuf Message, you're not relying on protobuf at all. No protobuf Class exists.

Reply:
> if can't load Protobuf Message class, the isProtobuf method can directly return false.

Yes